__kommandir__/bin/adept_openstack.py

```diff
@@ -0,0 +1,100 @@
+    raise RuntimeError("Script must be called as one of\n%s"
+                       % ALIASES.keys())
+
+
+MODE = ALIASES[BASENAME]
+if MODE == 'destroy':
+    if len(sys.argv) < 2:
+        raise RuntimeError("Must pass server name as first parameter")
+elif MODE == 'discover_create':
+    if len(sys.argv) < 3:
+        raise RuntimeError("Must pass server name, and one or more"
+                           " paths to public ssh key files as parameters")
+    PUB_KEY_FILES = sys.argv[2:]
+else:
+    raise RuntimeError("Unknown operation mode %s", MODE)
+NAME = sys.argv[1]
+
+
+# Assumed script running under an exclusive lock to prevent TOCTOU race
+session = os_client_config.make_rest_client('compute')
+response = session.get('/servers')
+servers_json = response.json()['servers']
+matched_server = [s for s in servers_json
+                  if s['name'] == NAME]
+
+
+if len(matched_server) > 1:
+    raise ValueError("Found more than one server named %s" % NAME)
+elif len(matched_server) == 1:
+    matched_server = matched_server[0]
+    response = session.get('/servers/%s' % matched_server['id'])
+    matched_server = response.json()['server']
+    MODE='discover'
+else:
+    MODE='create'
+    matched_server = None
+
+
+if MODE == 'discover':
+    networks = matched_server['addresses'].keys()
+    assert len(networks) == 1
+    network = networks[0]
+
+    ifaces = matched_server['addresses'][network]
+    floating_ipv4s = [iface['addr'] for iface in ifaces
+                      if iface['OS-EXT-IPS:type'] == 'floating']
+    assert len(floating_ipv4s) == 1
+    floating_ipv4 = floating_ipv4s[0]
+
+    # KISS
+    print "---"
+    print "inventory_hostname: %s" % NAME
+    print "ansible_host: %s" % floating_ipv4
+    print "ansible_ssh_host: {{ ansible_host }}"
+elif MODE == 'create':
+    pubkeys = []
+    for pub_key_file in PUB_KEY_FILES:
+        with open(pub_key_file,'rb') as key_file:
+            pubkeys.append(key_file.read())
+    auth_keys_json = str(pubkeys)
+
+    userdata = ("#cloud-config\n"
+                # Because I'm self-centered
+                "timezone: US/Eastern\n"
+                # Don't add silly 'please login as' to .ssh/authorized_keys
+                "disable_root: false\n"
+                # Allow password auth in case it's needed
+                "ssh_pwauth: True\n"
+                # Import all ssh_authorized_keys (below) into these users
+                "ssh_import_id: [root]\n"
+                # public keys to import to users (above)
+                "ssh_authorized_keys: %s\n" % auth_keys_json
+                # Prevent creating the default, generic user
+                "users:\n"
+                "   - name: root\n"
+                "     primary-group: root\n"
+                "     homdir: /root\n"
+                "     system: true\n")
+
+    # ref: https://developer.openstack.org/api-ref/compute/?expanded=create-server-detail#create-server
+else:  # mode == destroy
+    pass
```

Comment:
I'm fond of the python `if __main__` convention; it makes it easy to write unit tests.